templates/terraform/pre_delete/bigquery_dataset.go.erb

```diff
@@ -0,0 +1,8 @@
+// If we need to delete contents, we should customize the delete URL
+if d.Get("delete_contents_on_destroy").(bool) {
+	log.Printf("[DEBUG] replacing URL %q with a custom delete URL", url)
+	url, err = replaceVars(d, config, "<%= "{{#{object.__product.name}BasePath}}" -%><%=object.self_link-%>?deleteContents=true")
```

Comment:
Just curious- could you replace this code with overwriting the delete url in terraform.yaml to `projects/{{project}}/datasets/{{dataset_id}}?deleteContents={{delete_contents_on_destroy}}`?

Reply:
Yes we can, but we need to make `delete_contents_on_destroy` required (with a default of `false`). Which also means we need to do the hack with the import as the value is not returned by the API.

Reply:
The current resource has a DiffSuppressFunc to handle this, which would allow us to get rid of the custom code. If you add it to `ignore_read_extra` on the example, that'll take care of the `ImportStateVerifyIgnore` in the test.

Reply:
I'm still missing something. If I add back `DiffSuppressFunc: emptyOrDefaultStringSuppress("false")`, Terraform assumes that the `Default: false` can be replaced by an empty/null value, and will try to delete the dataset with a DELETE to, for example:
```
DELETE /bigquery/v2/projects/drebes-tf-test-c7e8/datasets/tf_test_ktxj7qqhou?alt=json&deleteContents= HTTP/1.1
```
which fails due to the missing `deleteContents`:
```
{
  "error": {
    "code": 400,
    "message": "Invalid value at 'delete_contents' (TYPE_BOOL), \"\"",
    "errors": [
      {
        "message": "Invalid value at 'delete_contents' (TYPE_BOOL), \"\"",
        "domain": "global",
        "reason": "badRequest"
      }
    ],
    "status": "INVALID_ARGUMENT"
  }
}
```

The only way I found that the deleteContents is explicitly passed as false is keeping the `Default: false`, marking it as `Required: true` and removing `DiffSuppressFunc: emptyOrDefaultStringSuppress("false")` (as you can't have a DiffSupressFunc for a required attribute.) Is there any way to have Terraform pass explicitly `false` as `defaultContents` whilie keeping the DiffSupressFunc and the field as Optional?

Reply:
It looks like Riley's been thinking about this too: https://github.com/GoogleCloudPlatform/magic-modules/pull/2213. Once that's merged, give it a try and see if making it a virtual field works.

Reply:
Will wait for #2213 to get merged and give it a try.

Reply:
Created the virtual field for delete_contents_on_destroy and it looks good. Doesn't seem to support a DiffSuppressFunc but I'm not sure we still need one.